packages/eventual-send/src/index.d.ts

```diff
@@ -1,61 +1,140 @@
+ *
+ * `Awaited<PromiseLike<T>>` recurses, and is just `T`.
+ * `Awaited<PromiseLike<PromiseLike<T>>>` is just `T` as well.
+ *
+ * @see {@link https://www.typescriptlang.org/docs/handbook/2/generics.html#handbook-content}
+ * @see {@link https://www.typescriptlang.org/docs/handbook/2/conditional-types.html}
+ */
 
-type ERef<T> = PromiseLike<T> | T;
+export type Callable = (...args: any[]) => any;
+
+/**
+// Same as https://github.com/microsoft/TypeScript/issues/31394
+ */
+export type ERef<T> = PromiseLike<T> | T;
+
+export declare const EmptyObj: {};
 
 // Type for an object that must only be invoked with E.  It supports a given
 // interface but declares all the functions as asyncable.
 export type EOnly<T> = T extends (...args: infer P) => infer R
-  ? (...args: P) => ERef<R> | EOnly<R>
-  : T extends Record<string | number | symbol, Function>
+  ? (...args: P) => ERef<Awaited<R>> | EOnly<Awaited<R>>
+  : T extends Record<PropertyKey, Callable>
   ? ERef<
       {
         [K in keyof T]: EOnly<T[K]>;
       }
     >
   : ERef<T>;
 
-type FilteredKeys<T, U> = {
+/**
+ * Return a union of property names/symbols/numbers P for which the record element T[P]'s type extends U.
+ *
+ * Given const x = { a: 123, b: 'hello', c: 42, 49: () => {}, 53: 67 },
+ *
+ * FilteredKeys<typeof x, number> is the type 'a' | 'c' | 53.
+ * FilteredKeys<typeof x, string> is the type 'b'.
+ * FilteredKeys<typeof x, 42 | 67> is the type 'c' | 53.
+ * FilteredKeys<typeof x, boolean> is the type never.
+ */
+export type FilteredKeys<T, U> = {
   [P in keyof T]: T[P] extends U ? P : never;
 }[keyof T];
 
-type DataOnly<T> = Omit<T, FilteredKeys<T, Function>>;
-type FunctionOnly<T> = Pick<T, FilteredKeys<T, Function>> &
-  (T extends Function ? (...args: Parameters<T>) => ReturnType<T> : {});
+/**
+ * `DataOnly<T>` means to return a record type `T2` consisting only of properties that are *not* functions.
+ */
+export type DataOnly<T> = Omit<T, FilteredKeys<T, Callable>>;
+
+// Nominal type to carry the local and remote interfaces of a Remotable.
+export declare class RemotableBrand<L, R> {
+  // The local properties of the object.
+  private localProperties: L;
 
-interface Remotable<T> {
-  __Remote__: T;
+  // The type of all the remotely-callable functions.
+  private remoteCallable: R;
 }
-type Remote<Primary, Local = DataOnly<Primary>> = ERef<
-  Local & Remotable<Primary>
+
+/**
+ * Creates a type that accepts both near and marshalled references that were
+ * returned from `Remotable` or `Far`, and also promises for such references.
+ */
+export type FarRef<Primary, Local = DataOnly<Primary>> = ERef<
+  Local & RemotableBrand<Local, Primary>
 >;
-type Unpromise<T> = T extends ERef<infer U> ? U : T;
 
-type Parameters<T> = T extends (...args: infer T) => any ? T : any;
-type ReturnType<T> = T extends (...args: any[]) => infer T ? T : any;
+/**
+ * `RemoteCallable<T>` means to return a record type `T2` consisting only of properties that are functions.
```

Comment:
What's this `T2` business? :) 
```suggestion
 * `RemoteCallable<T>` means to return a record type consisting only of properties that are functions.
```